### PR TITLE
Add support for multi service version testing to test framework

### PR DESCRIPTION
**Pull request by pakrym**
To enable multi-version testing add the `ClientTestFixture` attribute listing all the service versions to the test class itself or a base class:

```C#
[ClientTestFixture(
    BlobClientOptions.ServiceVersion.V2019_02_02,
    BlobClientOptions.ServiceVersion.V2019_07_07)]
public abstract class BlobTestBase : StorageTestBase
{
    private readonly BlobClientOptions.ServiceVersion _serviceVersion;

    public BlobTestBase(bool async, BlobClientOptions.ServiceVersion serviceVersion, RecordedTestMode? mode = null)
        : base(async, mode)
    {
        _serviceVersion = serviceVersion;
    }

    // ...
}
```

Add a service version attribute to the test class ctor and use the provided service version to create the `ClientOptions` instance.

```C#
public BlobClientOptions GetOptions() =>
    new BlobClientOptions(_serviceVersion) { /* ... */ };
```

To control what service versions test would run against use the `ServiceVersion` attribute by setting it's `Mix` or `Max` properties (inclusive).

```C#
[Test]
[ServiceVersion(Min = BlobClientOptions.ServiceVersion.V2019_02_02)]
public void Ctor_ConnectionString()
{
    // ...
}
```

How it looks it the test explorer:

image

### CODE REVIEW

File: sdk/storage/Azure.Storage.Blobs/tests/BlobTestBase.cs [hunk 1]
```diff
@@ -16,8 +16,10 @@
 
 namespace Azure.Storage.Test.Shared
 {
+    [ClientTestFixture(BlobClientOptions.ServiceVersion.V2019_02_02)]
```
Comment (pakrym) [inline sdk/storage/Azure.Storage.Blobs/tests/BlobTestBase.cs]: Add a second service version here when available. Follow the same pattern with other ***TestBase classes.

File: sdk/storage/Azure.Storage.Blobs/tests/BlobTestBase.cs [hunk 2]
```diff
@@ -28,11 +30,10 @@ public abstract class BlobTestBase : StorageTestBase
         protected string SecondaryStorageTenantSecondaryHost() =>
             new Uri(TestConfigSecondary.BlobServiceSecondaryEndpoint).Host;
 
-        public BlobTestBase(bool async) : this(async, null) { }
-
-        public BlobTestBase(bool async, RecordedTestMode? mode = null)
+        public BlobTestBase(bool async, BlobClientOptions.ServiceVersion serviceVersion, RecordedTestMode? mode = null)
             : base(async, mode)
         {
+            _serviceVersion = serviceVersion;
```
Comment (AlexanderSher) [inline sdk/storage/Azure.Storage.Blobs/tests/BlobTestBase.cs]: Can we access `serviceVersion` through the `TestContext.CurrentContext.Properties`?

File: sdk/core/Azure.Core/tests/TestFramework/RecordedTestBase.cs [hunk 3]
```diff
@@ -39,14 +39,15 @@ private string GetSessionFilePath(string name = null)
...
-        }
-
         [SetUp]
         public virtual void StartTestRecording()
         {
+            TestContext.TestAdapter test = TestContext.CurrentContext.Test;
+            if (Mode != RecordedTestMode.Live &&
+                test.Properties.ContainsKey("SkipRecordings"))
+            {
+                throw new IgnoreException((string) test.Properties.Get("SkipRecordings"));
```
Comment (tg-msft) [inline sdk/core/Azure.Core/tests/TestFramework/RecordedTestBase.cs]: Nit - add a comment like `// Only create test recordings for the latest version of the service`?

File: sdk/storage/Azure.Storage.Blobs/tests/BlobTestBase.cs [hunk 4]
```diff
@@ -45,7 +46,7 @@ public BlobTestBase(bool async, RecordedTestMode? mode = null)
 
         public BlobClientOptions GetOptions(bool parallelRange = false)
         {
-            var options = new BlobClientOptions
+            var options = new BlobClientOptions((BlobClientOptions.ServiceVersion) _serviceVersion)
```
Comment (tg-msft) [inline sdk/storage/Azure.Storage.Blobs/tests/BlobTestBase.cs]: Nit - we don't need the cast, do we?

**Review by tg-msft — APPROVED** [review]
This looks perfect!

File: sdk/storage/Azure.Storage.Blobs/tests/BlobTestBase.cs (hunk 4 above)
Comment (pakrym) [reply sdk/storage/Azure.Storage.Blobs/tests/BlobTestBase.cs]: Not anymore, leftovers from previous iteration.

**Comment by tg-msft** [comment]
There are a few other `*ClientOptions` that'll need this in the tests, but I filed #9587 to hunt them all down once this is merged.

/fyi @seanmcc-msft: we should enable this in the STG 71 branch once this PR is merged

**Review by JoshLove-msft — APPROVED** [review]

**Pull request merged**